swarm/network/stream/delivery.go

```diff
@@ -191,57 +171,84 @@ func (d *Delivery) Close() {
 	close(d.quit)
 }
 
-// RequestFromPeers sends a chunk retrieve request to a peer
-// The most eligible peer that hasn't already been sent to is chosen
-// TODO: define "eligible"
-func (d *Delivery) RequestFromPeers(ctx context.Context, req *network.Request) (*enode.ID, chan struct{}, error) {
-	requestFromPeersCount.Inc(1)
+// FindPeer is returning the closest peer from Kademlia that a chunk
+// request hasn't already been sent to
+func (d *Delivery) FindPeer(req *storage.Request) (*Peer, error) {
 	var sp *Peer
-	spID := req.Source
+	var err error
+
+	depth := d.kad.NeighbourhoodDepth()
+
+	d.kad.EachConn(req.Addr[:], 255, func(p *network.Peer, po int) bool {
+		id := p.ID()
+
+		// skip light nodes
+		if p.LightNode {
+			return true
+		}
 
-	if spID != nil {
-		sp = d.getPeer(*spID)
-		if sp == nil {
-			return nil, nil, fmt.Errorf("source peer %v not found", spID.String())
+		// do not send request back to peer who asked us. maybe merge with SkipPeer at some point
+		if req.Origin.String() == id.String() {
+			return true
 		}
-	} else {
-		d.kad.EachConn(req.Addr[:], 255, func(p *network.Peer, po int) bool {
-			id := p.ID()
-			if p.LightNode {
-				// skip light nodes
-				return true
-			}
-			if req.SkipPeer(id.String()) {
-				log.Trace("Delivery.RequestFromPeers: skip peer", "peer id", id)
-				return true
-			}
-			sp = d.getPeer(id)
-			// sp is nil, when we encounter a peer that is not registered for delivery, i.e. doesn't support the `stream` protocol
-			if sp == nil {
-				return true
-			}
-			spID = &id
+
+		// skip peers that we have already tried
+		if req.SkipPeer(id.String()) {
+			log.Trace("findpeer skip peer", "peer", id, "ref", req.Addr.String())
+			return true
+		}
+
+		// if origin is farther away from req.Addr and origin is not in our depth
+		prox := chunk.Proximity(req.Addr, d.kad.BaseAddr())
+		// proximity between the req.Addr and our base addr
+		if po < depth && prox >= depth {
+			log.Trace("findpeer skip peer because depth", "po", po, "depth", depth, "peer", id, "ref", req.Addr.String())
+
+			err = fmt.Errorf("not going outside of depth; ref=%s po=%v depth=%v prox=%v", req.Addr.String(), po, depth, prox)
 			return false
-		})
-		if sp == nil {
-			return nil, nil, errors.New("no peer found")
 		}
+
+		sp = d.getPeer(id)
+
+		// sp is nil, when we encounter a peer that is not registered for delivery, i.e. doesn't support the `stream` protocol
+		return sp == nil
```

Comment:
returning true means continue iterating. this is not obvious. in order to facilitate readability for the people who are not familiar with the core kademlia functionality, i would suggest something like this:
continueIterating := (sp == nil)
return continueIterating

Reply:
Sure, I can change that.